File: leetcode/solutions/0053.MaximumSubarray/maxSubArray.ts
export default function maxSubArray(nums: number[]): number {
  let max = nums[0];
  let prevMax = nums[0];

  for (let i = 1; i < nums.length; i++) {
    prevMax = Math.max(prevMax + nums[i], nums[i]);
    max = Math.max(max, prevMax);
  }

  return max;
}

File: leetcode/solutions/0053.MaximumSubarray/maxSubArray.go
package solution

func maxSubArray(nums []int) int {
	max, prevMax := nums[0], nums[0]

	for _, num := range nums[1:] {
		prevMax = maxInt(prevMax+num, num)
		max = maxInt(max, prevMax)
	}

	return max
}

func maxInt(num1, num2 int) int {
	if num1 > num2 {
		return num1
	}

	return num2
}
